### PR TITLE
Add Slab2 model to Rockhound

### LINKED ISSUE #61: Slab2 - A Comprehensive Subduction Zone Geometry Model
I would like to request the Slab2 model to be added to Rockhound.
This model contains the three-dimensional geometries of all active global subduction zones.
This kind of data is important to create realistic simulations of subduction zones in geodynamics and correct gravity data.

- Model/dataset name: Slab2
- Download link: https://www.sciencebase.gov/catalog/item/5aa1b00ee4b0b1c392e86467
- License: CC0 https://github.com/usgs/slab2/blob/master/LICENSE.md
- Target audience: geophysicists, solid earth, etc
- Is it widely used: yes

**Are you willing to help implement and maintain this feature?** Yes

### PR DESCRIPTION
**Pull request by aguspesce**
Add function to fetch and load data from Slab2: a geometry model of
subduction zones, provided by the [USGS](https://www.sciencebase.gov/catalog/item/5aa1b00ee4b0b1c392e86467). The model contains gridded data
for 27 subduction zones describing the depth, thickness, dip and strike
angles of the subducted slab. The data for each subduction zone is
fetched and loaded into a `xarray.Dataset`. The depth and thickness are
converted to meters and the coordinates renamed to "longitude" and
"latitude". Add metadata to the Dataset like the datum and the doi of
the model. Add netcdf4 as a requirement for Rockhound.

Fixes #61

**Reminders**

- [x] Run `make format` and `make check` to make sure the code follows the style guide.
- [x] Add tests for new features or tests that would have caught the bug that you're fixing.
- [x] Add new public functions/methods/classes to `doc/api/index.rst`.
- [x] Write detailed docstrings for all functions/methods.
- [x] If adding new functionality, add an example to the docstring, gallery, and/or tutorials.

### CODE REVIEW

**Comment by aguspesce** [comment]
@leouieda  and @santisoler,
I add in the registry file one by one the geometry files for each zone.
I do that because the USGS allows to download a zip file for a given zone but the hash in its name change every time that I download this file.

Do you have other idea to improve this problem?

**Comment by santisoler** [comment]
Happy to see this PR, @aguspesce!

I think it would be ideal to have a single zip file for each region on Slab2 model, but if it's not possible to have a consistent zip file in time, I think it's better to just add a ton of files to the registry.
I don't think this is a big issue. The registry file is not intended to be readable, so I don't see any problem with having a large number of files in it.

Ask for review whenever you are ready, or for help if you need it.

**Comment by aguspesce** [comment]
@santisoler  and @leouieda, 
The function to fetch Slab2 model  is almost finished but I need your help to do the testing functions.

File: examples/slab2.py [hunk 1]
```diff
@@ -0,0 +1,32 @@
...
+`USGS <https://www.sciencebase.gov/catalog/item/5aa1b00ee4b0b1c392e86467>`__ website.
+"""
+
+import rockhound as rh
+import matplotlib.pyplot as plt
+import cmocean
+import cartopy.crs as ccrs
+
+# Load a geometry of the Alaska subduction zone
+grid = rh.fetch_slab2(zone="south_america")
```
Comment (santisoler) [inline examples/slab2.py]: The zone in the comment does not corresponds with the actual fetched zone.

**Review by santisoler — CHANGES_REQUESTED** [review]
This is looking great @aguspesce! :1st_place_medal: 
Just found a little mistake on the example and left a few comments on the code, nothing to worry about.
I'll try to create some test functions to inspire you.

File: rockhound/slab2.py [hunk 2]
```diff
@@ -0,0 +1,141 @@
...
+    # Change long_name and units of longitude and latitude coords
+    ds.longitude.attrs["long_name"] = "Longitude"
+    ds.longitude.attrs["units"] = "degrees"
+    ds.latitude.attrs["long_name"] = "Latitude"
+    ds.latitude.attrs["units"] = "degrees"
+    # Add attributes to the xr.Dataset
+    ds.attrs.update(
+        {
+            "title": "Slab2 model - Zone: {}".format(ZONES[zone]["name"]),
+            "zone": ZONES[zone]["name"],
```
Comment (santisoler) [inline rockhound/slab2.py]: I think would be nice to have save the passed `zone` argument into the `zone` attribute of the `xr.Dataset`, but also is a good idea to save the full name of the zone.
So, maybe we could change this like:
```suggestion
            "zone": zone,
            "zone_full_name": ZONES[zone]["name"],
```

File: rockhound/slab2.py [hunk 3]
```diff
@@ -0,0 +1,141 @@
...
+            "datum": "WGS84",
+            "doi": "10.5066/F7PV6JNV.",
```
Comment (santisoler) [inline rockhound/slab2.py]: Remove the final dot:
```suggestion
            "doi": "10.5066/F7PV6JNV",
```

File: rockhound/slab2.py [hunk 4]
```diff
@@ -0,0 +1,141 @@
...
+        return fnames
+    arrays = [xr.open_dataarray(f).rename(x="longitude", y="latitude") for f in fnames]
+    for array, dataset in zip(arrays, DATASETS):
+        array.name = dataset
+        # Change long_name and add units of each array
+        array.attrs["long_name"] = DATASETS[dataset]["name"]
+        array.attrs["units"] = DATASETS[dataset]["units"]
+    # Merge arrays into a single xr.Dataset
+    ds = xr.merge(arrays)
+    # Change units of thickness, depth and depth_uncertainty to meters
```
Comment (santisoler) [inline rockhound/slab2.py:124]: We also need to modify the units of the `actual_range` attribute of the `thickness`, `depth` and `depth_uncertainty`. They are currently in the original units (km) what does not agree with the actual units of the data.

**Comment by santisoler** [comment]
I've just pushed a commit that adds test functions for Slab2. They will probably fail due to some differences between the expected values of the DOI and the `actual_range` attributes of `depth`, `thickness` and `depth_uncertainty` (see my previous comments). Don't worry, after they are fixed, the tests should pass.

In case you like my suggestion of adding the actual `zone` argument to the `zone` attribute, one test function must be updated.

File: rockhound/slab2.py (hunk 4 above)
Comment (aguspesce) [reply rockhound/slab2.py:124]: I resolved this!

File: examples/slab2.py (hunk 1 above)
Comment (aguspesce) [reply examples/slab2.py]: Done

**Comment by aguspesce** [comment]
Hello @santisoler, 
I make some changes... Place, see them

File: examples/slab2.py [hunk 5]
```diff
@@ -13,19 +13,48 @@
...
+    "philippines",
+    "puysegur",
+    "ryukyu",
+    "scotia_sea",
+    "solomon_islands",
+    "south_america",
+    "sulawesi",
+    "sumatra_java",
+    "vanuatu",
+}
```
Comment (santisoler) [inline examples/slab2.py]: There's no need to rewrite every zone available on Slab2.
You can import the `ZONES` dictionary directly and then iterate over its keys:
```python
from rockhound.slab2 import ZONES

...

for zone in ZONES:
    grid = rh.fetch_slab2(zone)
    ...
```

**Review by santisoler — CHANGES_REQUESTED** [review]
Just a suggestion to avoid repeating code.

File: examples/slab2.py (hunk 5 above)
Comment (aguspesce) [reply examples/slab2.py]: Done!

File: rockhound/slab2.py [hunk 6]
```diff
@@ -0,0 +1,147 @@
...
+            - `puysegur`: Puysegur
+            - `ryukyu`: Ryukyu
+            - `scotia_sea`: Scotia Sea
+            - `solomon_islands`: Solomon Islands
+            - `south_america`: South America
+            - `sulawesi`: Sulawesi
+            - `sumatra_java`: Sumatra-Java
+            - `vanuatu`: Vanuatu
+    load : bool
+        Wether to load the data into an :class:`xarray.Dataset` or just return the
```
Comment (santisoler) [inline rockhound/slab2.py]:
```suggestion
        Whether to load the data into an :class:`xarray.Dataset` or just return the
```

File: rockhound/slab2.py [hunk 7]
```diff
@@ -0,0 +1,147 @@
...
+            "zone_full_name": ZONES[zone]["name"],
+            "datum": "WGS84",
+            "doi": "10.5066/F7PV6JNV",
+        }
+    )
+    # Change unit of  actual_range attribute
+    grid.depth.attrs["actual_range"] = grid.depth.attrs["actual_range"] * 1000
+    grid.thickness.attrs["actual_range"] = grid.thickness.attrs["actual_range"] * 1000
+    grid.depth_uncertainty.attrs["actual_range"] = (
+        grid.depth_uncertainty.attrs["actual_range"] * 1000
```
Comment (santisoler) [inline rockhound/slab2.py]: This should be on the same place where the other units are changed, just to improve readability.

**Review by santisoler — CHANGES_REQUESTED** [review]

**Review by santisoler — APPROVED** [review]

File: doc/references.rst [hunk 8]
```diff
@@ -6,3 +6,4 @@ References
 .. [IRIS2011] IRIS DMC (2011), Data Services Products: EMC, A repository of Earth models, doi:10.17611/DP/EMC.1.
 .. [BEDMAP2] Fretwell, P. et al. (2013). Bedmap2: improved ice bed, surface and thickness datasets for Antarctica. The Cryosphere. doi:10.5194/tc-7-375-2013
 .. [Muller2008] Müller, R. D., Sdrolias, M., Gaina, C., & Roest, W. R. (2008). Age, spreading rates, and spreading asymmetry of the world’s ocean crust. Geochemistry, Geophysics, Geosystems, 9(4). doi:10.1029/2007GC001743
+.. [SLAB2] Hayes, G., 2018, Slab2 - A Comprehensive Subduction Zone Geometry Model: U.S. Geological Survey data release, https://doi.org/10.5066/F7PV6JNV.
```
Comment (santisoler) [inline doc/references.rst]:
```suggestion
.. [SLAB2] Hayes, G. (2018). Slab2 - A Comprehensive Subduction Zone Geometry Model: U.S. Geological Survey data release, https://doi.org/10.5066/F7PV6JNV.
```

File: examples/slab2.py [hunk 9]
```diff
@@ -0,0 +1,39 @@
...
+for zone in ZONES:
+    grid = rh.fetch_slab2(zone)
+
+    pc = grid.depth.plot.pcolormesh(
+        cmap=cmocean.cm.thermal_r,
+        ax=ax,
+        transform=ccrs.PlateCarree(),
+        add_colorbar=False,
+        vmin=-701160.33935547,
+        vmax=-1399.69277382,
```
Comment (santisoler) [inline examples/slab2.py]: Instead of hardcoding the min and max values I would compute them inside the example.

File: examples/slab2.py [hunk 10]
```diff
@@ -0,0 +1,39 @@
...
+    )
+
+ax.set_title("Slab2 - A comprehensive subduction zone geometry model")
```
Comment (santisoler) [inline examples/slab2.py]:
```suggestion
ax.set_title("Slab2: Geometry model for subduction zones")
```

**Review by santisoler — CHANGES_REQUESTED** [review]
This is looking great @aguspesce and I think we are very close to merge it :medal_sports:.

I would only improve some little things on the example.
Would you add some comments explaining what are you doing? The idea is that any new user could see a simple example on how this dataset could be loaded and plotted. So some comments are very welcome.

**Comment by santisoler** [comment]
This is looking awesome @aguspesce! Congratulations for the hard work!
I've just shrinked the colorbar on the example just to improve how the plot looks.
After the CIs pass I think this is ready to be merged.

**Review by santisoler — APPROVED** [review]

**Pull request merged**